crates/wasm-mutate/src/mutators/peephole/eggsy/mod.rs

```diff
@@ -0,0 +1,284 @@
+        (None, Some(_)) => Ordering::Greater,
+        (Some(_), None) => Ordering::Less,
+        (Some(a), Some(b)) => a.partial_cmp(&b).unwrap(),
+    }
+}
+
+impl<'a, CF, L, N> RandomExtractor<'a, CF, L, N>
+where
+    CF: CostFunction<L>,
+    L: Language,
+    N: Analysis<L, Data = Option<i32>>, // The analysis should return the index of the node in the e-class
+{
+    pub fn new(egraph: &'a EGraph<L, N>, cost_function: CF) -> Self {
+        let costs = HashMap::default();
+
+        let mut extractor = RandomExtractor {
+            costs,
+            egraph,
+            cost_function,
+        };
+        extractor.costs = extractor.find_costs();
+        extractor
+    }
+
+    fn find_costs(&mut self) -> HashMap<Id, (CF::Cost, usize)> {
+        let mut costs = HashMap::new();
+
+        let mut did_something = true;
+        while did_something {
+            did_something = false;
+
+            for class in self.egraph.classes() {
+                let pass = self.make_pass(&mut costs, class);
+                match (costs.get(&class.id), pass) {
+                    (None, Some(new)) => {
+                        costs.insert(class.id, new);
+                        did_something = true;
+                    }
+                    (Some(old), Some(new)) if new.0 < old.0 => {
+                        costs.insert(class.id, new);
+                        did_something = true;
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        costs
+    }
+
+    fn make_pass(
+        &mut self,
+        costs: &mut HashMap<Id, (CF::Cost, usize)>,
+        eclass: &EClass<L, Option<i32>>,
+    ) -> Option<(CF::Cost, usize)> {
+        let (cost, node_idx) = eclass
+            .iter()
+            .enumerate()
+            .map(|(i, n)| (self.node_total_cost(n, costs), i))
+            .min_by(|a, b| cmp(&a.0, &b.0))
+            .unwrap_or_else(|| panic!("Can't extract, eclass is empty: {:#?}", eclass));
+        cost.map(|c| (c, node_idx))
+    }
+
+    fn node_total_cost(
+        &mut self,
+        node: &L,
+        costs: &mut HashMap<Id, (CF::Cost, usize)>,
+    ) -> Option<CF::Cost> {
+        let egraph = self.egraph;
+        let has_cost = |&id| costs.contains_key(&egraph.find(id));
+        if node.children().iter().all(has_cost) {
+            let costs = &costs;
+            let cost_f = |id| costs[&egraph.find(id)].0.clone();
+            Some(self.cost_function.cost(&node, cost_f))
+        } else {
+            None
+        }
+    }
+
+    pub fn get_costs(&self) -> &HashMap<Id, (<CF as CostFunction<L>>::Cost, usize)> {
+        &self.costs
+    }
+
+    /// Here is where the random magic is done
+    pub fn generate_random_tree(
+        &self,
+        rnd: &mut rand::prelude::SmallRng,
+        eclass: Id,
+        max_depth: u32,
+    ) -> crate::Result<(Vec<&L>, Vec<Vec<Id>>)> // return the random tree, TODO, improve the way the tree is returned
```

Comment:
Ah so the idea is that we don't actually need to construct the `RecExpr` here since it is just an intermediate state on our way to Wasm and we can skip it and go directly to Wasm?

Reply:
Yep, thats correct